File: src/backend/kcvstore.py
import sqlite3

class KCVStore:
    def __init__(self, dbpath):
        self.dbpath = dbpath
        self.conn = sqlite3.connect(self.dbpath)
        self.cursor = self.conn.cursor()
        self.cursor.execute('''
            CREATE TABLE IF NOT EXISTS kcvstore (
                key TEXT PRIMARY KEY,
                category TEXT,
                value TEXT
            )
        ''')
        self.conn.commit()

    def set(self, key, category, value):
        self.cursor.execute('''
            INSERT OR REPLACE INTO kcvstore (key, category, value)
            VALUES (?, ?, ?)
        ''', (key, category,value))
        self.conn.commit()

    def get(self, key, category):
        self.cursor.execute('''
            SELECT value FROM kcvstore WHERE key = ? and category = ?
        ''', (key,category))
        row = self.cursor.fetchone()
        return row[0] if row else None

    def delete(self, key,category):
        self.cursor.execute('''
            DELETE FROM kcvstore WHERE key = ? and category = ?
        ''', (key,category))
        self.conn.commit()

    def close(self):
        self.conn.close()

File: src/csbackend/Agents/KCVStore.cs
namespace agents;
using Microsoft.Data.Sqlite;

public class KCVStore
{
    public string CONNECTION_STRING { get; set; } = null!;

    public KCVStore(string connStr)
    {
        CONNECTION_STRING = connStr;

        // Create table if not exists
        using var conn = new SqliteConnection(connStr);
        conn.Open();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = @"
            CREATE TABLE IF NOT EXISTS kcvstore (
                key TEXT NOT NULL,
                category TEXT NOT NULL,
                value TEXT NOT NULL,
                PRIMARY KEY (key, category)
            )
        ";
        cmd.ExecuteNonQuery();
    }

    public async Task<string> Get(string key, string category)
    {
        using var conn = new SqliteConnection(CONNECTION_STRING);
        await conn.OpenAsync();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = @"
            SELECT value FROM kcvstore WHERE key = @key AND category = @category
        ";
        cmd.Parameters.AddWithValue("@key", key);
        cmd.Parameters.AddWithValue("@category", category);
        var row = cmd.ExecuteScalar();
        return row?.ToString() ?? string.Empty;
    }

    public async Task Set(string key, string category, string value)
    {
        using var conn = new SqliteConnection(CONNECTION_STRING);
        await conn.OpenAsync();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = @"
            INSERT OR REPLACE INTO kcvstore (key, category, value)
            VALUES (@key, @category, @value)
        ";
        cmd.Parameters.AddWithValue("@key", key);
        cmd.Parameters.AddWithValue("@category", category);
        cmd.Parameters.AddWithValue("@value", value);
        cmd.ExecuteNonQuery();
    }

    public async Task Delete(string key, string category)
    {
        using var conn = new SqliteConnection(CONNECTION_STRING);
        await conn.OpenAsync();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = @"
            DELETE FROM kcvstore WHERE key = @key AND category = @category
        ";
        cmd.Parameters.AddWithValue("@key", key);
        cmd.Parameters.AddWithValue("@category", category);
        cmd.ExecuteNonQuery();
    }
}
